discovery_engine_core/core/src/storage/sqlite/utils.rs

```diff
@@ -39,3 +43,75 @@ where
         self
     }
 }
+
+pub(crate) trait SqlxSqliteErrorExt<V> {
+    /// Use this on the result of a sqlite query which might have failed a foreign
+    /// key constraint when an illegal document was passed in.
+    ///
+    /// For example it can be used in `.user_reacted` to handle it being
+    /// called with a random document id.
+    ///
+    /// If there is an error and it is a fk violation an appropriate error variant is
+    /// used instead of the default generic database error.
+    fn fk_violation_is_invalid_document_id(self, id: document::Id) -> Result<V, Error>;
```

Comment:
❓ should we also use the `on_` naming since this method acts on an error as well?
```suggestion
    fn on_fk_violation_is_invalid_document_id(self, id: document::Id) -> Result<V, Error>;
```